app/controllers/admin/badges_controller.rb

```diff
@@ -52,37 +54,79 @@ def mass_award
     end
 
     replace_badge_owners = params[:replace_badge_owners] == 'true'
-    BadgeGranter.revoke_all(badge) if replace_badge_owners
+    grant_existing_holders = params[:grant_existing_holders] == 'true'
+    if grant_existing_holders && !badge.multiple_grant?
+      render_json_error(
+        I18n.t('badges.mass_award.errors.cant_grant_multiple_times', badge_name: badge.display_name),
+        status: 422
+      )
+      return
+    end
 
-    batch_number = 1
     line_number = 1
-    batch = []
-
+    entries = []
     File.open(csv_file) do |csv|
-      mode = Email.is_valid?(CSV.parse_line(csv.first).first) ? 'email' : 'username'
-      csv.rewind
+      csv.each_line do |line|
+        line = CSV.parse_line(line).first&.strip&.downcase
+        line_number += 1
 
-      csv.each_line do |email_line|
-        line = CSV.parse_line(email_line).first
+        entries << line if line.present?
 
-        if line.present?
-          batch << line
-          line_number += 1
+        if entries.size > MAX_CSV_LINES
+          return render_json_error I18n.t('badges.mass_award.errors.too_many_csv_entries', count: MAX_CSV_LINES), status: 400
         end
+      end
+    end
 
-        # Split the emails in batches of 200 elements.
-        full_batch = csv.lineno % (BadgeGranter::MAX_ITEMS_FOR_DELTA * batch_number) == 0
-        last_batch_item = full_batch || csv.eof?
+    BadgeGranter.revoke_all(badge) if replace_badge_owners
 
-        if last_batch_item
-          Jobs.enqueue(:mass_award_badge, users_batch: batch, badge_id: badge.id, mode: mode)
-          batch = []
-          batch_number += 1
-        end
+    usernames = []
+    emails = []
+    entries.each do |entry|
+      if entry.include?('@')
+        emails << entry
+      else
+        usernames << entry
       end
     end
 
-    head :ok
+    usernames_map_to_ids = emails_map_to_ids = {}
+    if usernames.size > 0
+      usernames_map_to_ids = User.where(username_lower: usernames).pluck(:username_lower, :id).to_h
+    end
+    if emails.size > 0
+      emails_map_to_ids = User.with_email(emails).pluck('LOWER(user_emails.email)', :id).to_h
+    end
+
+    count_per_user = {}
+    unmatched = Set.new
+    entries.each do |entry|
+      id = usernames_map_to_ids[entry] || emails_map_to_ids[entry]
+      if id.blank?
+        unmatched << entry
+        next
+      end
+      if grant_existing_holders
+        count_per_user[id] ||= 0
+        count_per_user[id] += 1
+      else
+        count_per_user[id] = 1
+      end
+    end
+    count_per_user.each do |user_id, count|
+      Jobs.enqueue(
+        :mass_award_badge,
+        user: user_id,
+        badge: badge.id,
+        count: count,
+        grant_existing_holders: grant_existing_holders
+      )
+    end
+
+    render json: {
+      unmatched_entries: unmatched.to_a,
```

Comment:
Hmm I think we need to be careful here. If they accidentally upload a wrong CSV in the right format, `unmatched` might be an array containing 50K items.